products/batch_exports/backend/temporal/metrics.py

```diff
@@ -375,3 +341,82 @@ def get_interval_from_bounds(
                 interval = f"every_{int(s / 60)}_minutes"
 
     return interval
+
+
+def get_sla_from_interval(
+    interval: str,
+) -> dt.timedelta:
+    """Get the SLA for a batch export based on its interval string."""
+    match interval:
+        case "hour":
+            return dt.timedelta(hours=1)
+        case "day":
+            return dt.timedelta(days=1)
+        case "week":
+            return dt.timedelta(days=7)
+        case interval:
+            _, value, unit = interval.split("_")
+            kwargs = {unit: int(value)}
+            return dt.timedelta(**kwargs)
+
+
+class SLAWaiter:
+    """Wait until a batch export has exceeded SLA and log a warning.
+
+    Attributes:
+        name: The name of the task we are waiting for. Will be used in the log
+            message if SLA is exceeded.
+        sla: The SLA we are waiting for.
+
+    Examples:
+        Nothing happens when no SLA is exceeded.
+
+        >>> async with SLAWaiter(name="my-task", sla=dt.datetime(seconds=10)) as waiter:
+        ...     await asyncio.sleep(1)
+        ...     waiter.is_over_sla()
+        False
+
+        A log will be printed if SLA is exceeded.
+
+        >>> async with SLAWaiter(name="my-task", sla=dt.datetime(seconds=1)) as waiter:
+        ...     await asyncio.sleep(10)
+        ...     waiter.is_over_sla()
+        True
+    """
+
+    def __init__(self, name: str, sla: dt.timedelta):
+        self.name = name
+        self.sla = sla
+        self._over_sla = asyncio.Event()
+        self._waiter: asyncio.Task[None] | None = None
+
+    def is_over_sla(self) -> bool:
+        return self._over_sla.is_set()
+
+    async def __aenter__(self) -> typing.Self:
+        self._waiter = asyncio.create_task(self.wait_for_sla())
+        return self
+
+    async def __aexit__(self, exc_type: type[BaseException] | None, exc_value: BaseException | None, traceback) -> None:
+        """Reset internal state and cancel waiter."""
+
+        if self._waiter:
+            is_running = self._waiter.cancel()
+
+            if is_running:
+                _ = await asyncio.wait([self._waiter])
+
+            self._waiter = None
+
+        self._over_sla.clear()
+
+    async def wait_for_sla(self) -> None:
+        """Coroutine used to wait for SLA seconds."""
+        await asyncio.sleep(self.sla.total_seconds())
+
+        self._over_sla.set()
+        LOGGER.warning(
+            "%(name)s has been running longer than SLA of %(sla_seconds)ds",
```

Comment:
think this is fine for now but if we add more of these types of logs, we may want to add a standardized prefix so we can filter for them in a dashboard more easily